std/algorithm/comparison.d

```diff
@@ -1911,3 +1915,97 @@ bool isPermutation(alias pred = "a == b", Range1, Range2)
         [mytuple(2, 3), mytuple(1, 2)]
     ));
 }
+
+enum implicitlyConvertsToBool(T) = isImplicitlyConvertible!(bool, T) || isArray!T;
+
+/** Returns: First Parameter (element of $(D a)) whose conversion to $(D bool)
+    is $(D true). If all arguments $(D bool)-convert to $(D false) return last
+    parameter.
+
+    Similar to behaviour of `or` operator in dynamic languages such as Lisp's
+    `(or a b ...)` and Python's `a or b or ...` except that last argument is
+    returned if no argument could be implicitly converted to $(D true).
+
+    NOTE: Lazy parameters are currently, too restrictively, inferred by DMD to
+    always throw eventhough they don't need to be. This makes it impossible to
+    mark $(D either) as $(D nothrow). Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(alias pred = a => a, Ts...)(lazy Ts a)
+if (a.length > 1 &&
+    !is(CommonType!Ts == void) &&
+    allSatisfy!(implicitlyConvertsToBool, Ts))
```

Comment:
This condition is wrong because what you need is to make sure that `pred(x)` is if-testable, not `x`.